x/staking/handler.go

```diff
@@ -229,7 +229,17 @@ func handleMsgDelegate(ctx sdk.Context, msg types.MsgDelegate, k keeper.Keeper)
 }
 
 func handleMsgUndelegate(ctx sdk.Context, msg types.MsgUndelegate, k keeper.Keeper) sdk.Result {
-	completionTime, err := k.Undelegate(ctx, msg.DelegatorAddress, msg.ValidatorAddress, msg.SharesAmount)
+	validator, found := k.GetValidator(ctx, msg.ValidatorAddress)
+	if !found {
+		return ErrNoValidatorFound(k.Codespace()).Result()
+	}
+
+	if validator.Tokens.IsZero() {
+		return ErrInsufficientShares(k.Codespace()).Result()
+	}
+
+	sharesAmt := validator.DelegatorShares.MulInt(msg.Amount.Amount).QuoInt(validator.Tokens).TruncateDec()
```

Comment:
Lastly, it may make sense to take the cap of the sharesAmt here - aka if the message's sharesAmt > Delegations.Shares then just use the delegations shares amount... This could occur due to rounding, however we also don't want to just truncate the shares or take the minimum because we want to allow for the full withdraw of shares from a delegation.

Reply:
ACKed.